Add ActiveStorage Instrumentation

cwoodcox opened the pull request:
An updated version of #1290, this adds ActiveStorage instrumentation via `ActiveSupport::Notifications` just like the other Rails integrations.

# To-do

- [x] add `@public_api` tags to the proper methods/modules to match the rest
- [x] add documentation comments for the individual event modules
- [x] change span_name in all events to be the event name (delete, exist, etc)
- [x] change resource name to be blob name
  - include storage service name?
- [ ] add tests

ericmustin committed:
add initial active storage integration skeleton

moved and re-namespaced everything

cwoodcox committed:
refactor everything to match the new class hierarchy

cwoodcox committed:
remove old ENV references and default service name

cwoodcox committed:
update comments and add @public_api comments

cwoodcox committed:
change span name to reflect the event name

the event name is the actual name of the operation performed

cwoodcox committed:
use service and key as resource name

cwoodcox committed:
only set service name if it is configured

marcotc commented on lib/datadog/tracing/contrib/active_storage/events/upload.rb:
```diff
@@ -0,0 +1,58 @@
+              # checksum is also on payload but this has infinite cardinality and low value
+
+              span.service = configuration[:service_name] if configuration[:service_name]
+              span.resource = "#{as_service}: #{as_key}"
```
I believe `as_key` has high cardinality, as it seems to be [unique per file/path stored](https://github.com/rails/rails/blob/5a5440ba727b5b9882ee49763f8d44ed4ffde017/activestorage/app/controllers/active_storage/direct_uploads_controller.rb#L8).
This falls under the same unresolved issue we have with instrumenting URL paths for arbitrary URLs: there's no reliable default quantizer we can think of that wouldn't cause high cardinality for some users.

I recommend the removal of `as_key` here, and instead set it as a tag (`span.set_tag(...,as_key)`).

Please let me know if my assumptions are incorrect and `as_key` is either a predictable value or can be quantized reliably.

marcotc reviewed (commented):
This is an awesome PR, super flushed out already, thank you @cwoodcox! 🙇 
With small tweaks and testing this can be merged very soon.

I left an inline comment for us to discuss.

Besides testing, one thing missing is user-facing documentation, which goes in the https://github.com/DataDog/dd-trace-rb/blob/master/docs/GettingStarted.md file. You can certainly leave it until the end to add it though.

ivoanjo commented:
This PR has been open for quite a while; @cwoodcox is this something you're still interested in pushing forward?

cwoodcox commented:
I'll dig into this again next month, I think it would be good :)

Commit:
add frozen string literal decls

Commit:
remove typed declarations

Commit:
use require_relative

Commit:
typo fix

Commit:
test suite by claude

Commit:
rake task

Commit:
tag component & operations

Commit:
assert component and operation tags

Commit:
missing tests

Commit:
fix configuration

Commit:
getting tests to run

Commit:
getting tests to run

Commit:
span_type -> span.type

Commit:
configurations

Commit:
configurations

Commit:
trailing commas

Commit:
dry

Commit:
dry tests

vpellan committed:
Fix spec issues

vpellan committed:
Fix more specs

vpellan committed:
standard and steep

vpellan committed:
Revert rails6 app changes

Strech reviewed (commented):
Great job 👏🏼 

I would like to suggest improvements over the typing to:

1. Remove values from the use in String literals as it makes things fragile
2. Do not use steep:ignore in areas with issues
3. Refer 3rd party types with `::`-prefix

p-ddsign committed:
Apply standardrb formatting to ActiveStorage integration

p-ddsign committed:
Use storage service name as ActiveStorage span resource

The object key/prefix is unique per stored file and produces unbounded
resource cardinality. The key remains available as a span tag.

p-ddsign committed:
Address ActiveStorage typing review feedback

- Fix compatible? to handle nil version without steep:ignore
- Use unqualified Patcher reference
- Type string constants and event names as ::String
- Generalize span_options and fully-qualify the Subscription type

p-ddsign committed:
Instrument ActiveStorage from the Rails integration

ActiveStorage#auto_instrument? is false under Rails, so the Rails
framework setup must activate it explicitly, matching the other Rails
sub-integrations.

p-ddsign committed:
Document ActiveStorage integration in GettingStarted
